oyjl.1
.TH oyjl 1 "November 12, 2017" "User Commands"
.SH NAME
oyjl \- Light weight JSON parse and manipulation tool


.SH SYNOPSIS
\fBoyjl\fR [-j|-c|-k|-p|-s \fISTRING\fR] [-v] [-i \fIFILENAME\fR] [-x \fIPATH\fR]


.SH DESCRIPTION
The oyjl program can be used to parse, filter sub trees, select values and modify values in JSON texts.


.SH OPTIONS
.TP
.B \-j\fR Print JSON to std output.
This option requires a -i JSON FILENAME and a -x XPATH.

.TP
.B \-c\fR Print count of leaves in node.
This option requires a -i JSON FILENAME.

.TP
.B \-k\fR Print key name of node.
This option requires a -i JSON FILENAME.

.TP
.B \-p\fR Print all matching paths.
This option requires a -i JSON FILENAME and a -x XPATH.

.TP
.B \-s\fR Set a key name to a value.
This option requires a -i JSON FILENAME, a -x XPATH and a -s STRING.


.SH GENERAL OPTIONS
.TP
.B \-i\fR \fIFILENAME\fR, \-i\fR=\fIFILENAME\fR a file or stream
A JSON file name or a input stream like "stdin".

.TP
.B \-x\fR \fIXPATH\fR, \-x\fR=\fIXPATH\fR path specifier.
The path consists of slash '/' separated terms. Each term can be a key name
or a square bracketed index. A empty term is used for a search inside a tree.

.TP
.B \-v, \-\-verbose\fR increase verbosity


.SH EXAMPLES
.TP
Print a key value or sub tree:
.B oyjl -i text.json -x ///[0]

.TP
Print the number of keys:
.B oyjl -c -i text.json -x my/path/

.TP
Print a key name:
.B oyjl -k -i text.json -x ///[0]

.TP
Print all matching xpaths:
.B oyjl -p -i text.json -x //

.TP
Modify a key/value:
.B oyjl -i text.json -x my/path/to/key -s value


.SH AUTHOR
Kai-Uwe Behrmann (ku.b (at) gmx.de)

  
.SH COPYRIGHT
(c) 2017, Kai-Uwe Behrmann and others
.fi
License: MIT <http://www.opensource.org/licenses/MIT>

  
.SH "SEE ALSO"
https://codedocs.xyz/oyranos-cms/oyranos/group__oyjl.html


.SH "BUGS"
at: https://github.com/oyranos-cms/oyranos/issues


ChangeLog
Version 0.9.5

Andreas Sturmlechner (3):
	* [conf]: fix CUPS macro names
	* [core]: more typo and grammar fixes
	* [core]: Typo fix: s/Promt/Prompt/

Kai-Uwe Behrmann (349):
	* [core]: fix compile warning in oy_io_core
	* [core]: remove unused internal Path configuration APIs
	* [core]: fix array size
	* [docu]: fix typo
	* [core]: remove unused oySetDefaultProfileBlock()
	* [core]: skip redundant oy_domain_codeset declaration
	* [core]: use braces around empty body in debug macros
	* [core]: remove duplicate declaration for RegMatch
	* [tools]: fix -pedantic compile warnings
	* [core]: fix -pedantic compile warnings
	* [tools]: check ICC ID in oy-profile -v
	* [sources]: support ICC ID in oyProfile_FromFile()
	* [alpha]: remove double defines
	* [API]: avoid OY_COMPUTE flags ambiguity
	* [docu]: add cmake flags section to README
	* [sources]: add comments to OY_COMPUTE define
	* [sources]: error messages only for wrong hash
	* [core]: improve hash error message
	* [tools]: adapt to oyConversion_CreateBasicPixelsFromBuffers()
	* [core]: add more warnings to profile ID check
	* [lcm2]: fix run time warnings
	* [conf]: use Libs.private tag in oyranos.pc
	* [qarz]: return sRGB for non colour server profile
	* [core]: fix wrong free of default profile name
	* [lcm2]: use oyLibNameCreate_ to support osX
	* [core]: add internal oyLibNameCreate_()
	* [sources]: add warning msg to cmm_info->init()
	* [exmpl]: compile fix under osX
	* [oyX1]: fix osX build
	* [oydi]: fix color server activity detection
	* [lcm2]: use lib names from cmake for dlopen
	* [sources]: skip modules with erroneous init state
	* [tools]: remove lcms2 header from oy-icc
	* [oforms]: skip alpha header
	* [core]: let cmake detect library prefix and suffix
	* [sources]: fix compile warning in release mode
	* [build]: sync RPM spec file with OBS
	* [core]: remove too new attribute
	* [core]: guard pthread_mutexattr_setrobust()
	* [conf]: remove unused ELEKTRA_LIBS variable
	* [conf]: relax about Elektra 0.7 spec requirement
	* [conf]: remove ELEKTRA_LIBS from normal flags
	* [conf]: support elektra-static in oy-config
	* [tools]: fix compiler warning in oy-icc
	* [i18n]: update Deutsch
	* [tools]: make more strings translateable
	* [test]: oyConversion_CreateBasicPixelsFromBuffers() API
	* [sources]: use pixel types in oyConversion_CreateBasicPixelsFromBuffers() II
	* [lcm2]: fix OpenMP threading issue for proofing
	* [lcm2]: support lcms-2.6 cmsContext changes
	* [lcm2]: add cmsContext function pointers
	* [exmpl]: support -f fallback-icc in oy-test-device and oy-file
	* [oyRE]: generate contexts and fallback profile
	* [lraw]: remove matrix profile generation from raw-image node
	* [docu]: pass all public headers to doxygen
	* [oyX1]: let icc_profile.fallback be explicite
	* [exmpl]: use "properties" backend call for verbose oy-test-device -l
	* [oyRE]: route Configs_FromPattern() through Configs_Modify()
	* [exmpl]: allow --device-name in oy-test-device -l
	* [docu]: disambiguate description sentence in oyX1
	* [exmpl]: show "supported_devices_info" in oy-test-device -v
	* [exmpl]: support --device-name option in oy-test-device
	* [oyRE]: rename message variable
	* [docu]: describe the oy-file --profile option
	* [lraw]: write meaningful profile name
	* [core]: use posix signal condition with empty workload
	* [sources]: unlock previously locked objects
	* [sources]: init oyObject_s::_lock
	* [oyRE]: set new module_api after API break
	* [docu]: fix fragment markup for newer doxygen
	* [source]: use pixel types in oyConversion_CreateBasicPixelsFromBuffers()
	* [core]: init mutex attribute
	* [core]: guard global objetc counter
	* [core]: improve thread communication
	* [test]: simplify graph call
	* [docu]: describe basic function types
	* [core]: skip allocation when file size is zero
	* [sources]: release oyFilterNode_ToBlob() blob if not set
	* [tools]: sync oy-icc -f icc with other formats -n option
	* [sources]: move oyFilterGraph_ToBlob to FilterNode
	* [exmpl]: keep image_display responsive while editing ICC options
	* [test]: sync test_oy with test2 arguments
	* [tools]: simplify graph call in oy-icc
	* [tools]: oy-profile -lvn reduce prints to stdout
	* [test]: increase fuzziness for Lab conversion
	* [sources]: fix oyPointer_Set() release object ptr
	* [lcms]: dlopen lcms
	* [sources]: fix display of locale info in ICC text tags
	* [tools]: move debug messages to stderr in oy-monitor
	* [sources]: skip repeat context creation in oyConversion_RunPixels()
	* [core]: internally expose oyReadFileSize_()
	* [core]: fix printf formating for hashes
	* [core]: support oyGetTempFileName_( NULL ) file name
	* [core]: add internal oyGetTempFileName_()
	* [docu]: add doxygen mainpage page name
	* [templates]: set const oyStruct_s::type_ member
	* [docu]: describe all oyStruct_s members
	* [docu]: describe per filter context creation and caching
	* [docu]: describe oyCMMapi6_GetConvertF
	* [docu]: add object sources to internal Doxyfile
	* [docu]: add brief doxygen group description
	* [sources]: add oyFilterCore_S/GetBackendContext()
	* [test]: select test by number
	* [sources]: pass context in oyCMMuiGet_f definition
	* [templates]: use type range check in oyCMMapiFilter derived object types
	* [templates]: add oyCheckTypeRange_m macro
	* [API]: reorder object types
	* [sources]: add oyCMMui::parent member
	* [core]: improve oyLibPathsGet_()
	* [tools]: init OY_DEBUG in oy-xforms-modules
	* [sources]: support oySIGNATURE_ATTRIBUTES in oyProfile_SetSignature()
	* [sources]: fix wrong pointer casting
	* [sources]: fix compiler warning in Hash object
	* [templates]: fix modules listing in oy-xforms-moduls -l
	* [build]: disable cmml
	* [core]: fix compile warnings
	* [conf]: fix oy-config script libraries
	* [API]: add FilterNode_S/Get[Backend]Context()
	* [template]: add CMMapiFilter:runtime_context
	* [docu]: describe some Pointer function types
	* [docu]: add modules declarations
	* [docu]: fix typos
	* [docu]: fix typo

oyjl/oyjl.c
/*  @file oyjl.c
 *
 *  oyjl - Yajl tree extension
 *
 *  @par Copyright:
 *            2016-2018 (C) Kai-Uwe Behrmann
 *
 *  @brief    Oyjl command line
 *  @internal
 *  @author   Kai-Uwe Behrmann <ku.b@gmx.de>
 *  @par License:
 *            MIT <http://www.opensource.org/licenses/mit-license.php>
 *  @since    2016/12/17
 */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>

#include "oyjl_tree_internal.h"
#include "oyjl_macros.h"


void printfHelp(int argc OYJL_UNUSED, char ** argv)
{
  if(strstr(argv[0],"jsontoyaml"))
  {
  fprintf( stderr, "\n");
  fprintf( stderr, "%s %s\n",   argv[0],
                                _("is a JSON to YAML converter"));
  fprintf( stderr, "\n");
  fprintf( stderr, "%s\n",                 _("Usage"));
  fprintf( stderr, "      %s FILE_NAME\n",        argv[0]);
  fprintf( stderr, "\n");
  fprintf( stderr, "\n");
    return;
  }


  fprintf( stderr, "\n");
  fprintf( stderr, "%s %s\n",   argv[0],
                                _("is a JSON parse tool"));
  fprintf( stderr, "\n");
  fprintf( stderr, "%s\n",                 _("Usage"));
  fprintf( stderr, "  %s\n",               _("Print:"));
  fprintf( stderr, "      %s [-j|-c|-k|-p|-s STRING] [-v] [-i FILE_NAME] [-x PATH]\n",        argv[0]);
  fprintf( stderr, "        -j\tprint JSON - default mode\n");
  fprintf( stderr, "        -y\tprint YAML - better human readable\n");
  fprintf( stderr, "        -c\tprint node count\n");
  fprintf( stderr, "        -k\tprint key name\n");
  fprintf( stderr, "        -p\tprint all matching paths\n");
  fprintf( stderr, "        -v\tset a key to a value\n");
  fprintf( stderr, "        -i FILE_NAME  %s\n", _("JSON file"));
  fprintf( stderr, "        -x PATH\tpath expression, indexing is in edged brackets \"[NUMBER]\"\n");
  fprintf( stderr, "\n");
  fprintf( stderr, "  %s\n",               _("Print a help text:"));
  fprintf( stderr, "      %s -h\n",        argv[0]);
  fprintf( stderr, "\n");
  fprintf( stderr, "  %s\n",               _("General options:"));
  fprintf( stderr, "        -v              %s\n", _("verbose"));
  fprintf( stderr, "\n");
  fprintf( stderr, "%s\n",                 _("Example"));
  fprintf( stderr, "  %s\n",               _("Print JSON:"));
  fprintf( stderr, "      %s -j -i example.json -x [0]\n",        argv[0]);
  fprintf( stderr, "\n");
  fprintf( stderr, "  %s\n",               _("Print count:"));
  fprintf( stderr, "      %s -c -i example.json -x [0]/[0]\n",        argv[0]);
  fprintf( stderr, "\n");
  fprintf( stderr, "  %s\n",               _("Print key name:"));
  fprintf( stderr, "      %s -k -i example.json -x [0]/[0]\n",        argv[0]);
  fprintf( stderr, "\n");
  fprintf( stderr, "  %s\n",               _("Print path:"));
  fprintf( stderr, "      %s -p -i example.json -x /\n",        argv[0]);
  fprintf( stderr, "\n");
  fprintf( stderr, "  %s\n",               _("Set key/value:"));
  fprintf( stderr, "      %s -i example.json -x [0]/[0]/key -s \"value\"\n",        argv[0]);
  fprintf( stderr, "\n");
  fprintf( stderr, "\n");
}


typedef enum {
  JSON,
  YAML,
  COUNT,
  TYPE,
  KEY,
  PATHS
} SHOW;

int main(int argc, char ** argv)
{
  SHOW show = JSON;
  int verbose = 0;
  int error = 0;
  int size = 0;
  char * text = NULL;
  const char * value_string = NULL;
  const char * input_file_name = NULL,
             * xpath = NULL;
  oyjl_val root = NULL,
           value = NULL;
  int index = 0;

#ifdef USE_GETTEXT
  setlocale(LC_ALL,"");
#endif

  if(strstr(argv[0],"jsontoyaml"))
  {
    if(argc > 1)
      input_file_name = argv[1];
    else
    {
      printfHelp(argc, argv);
      exit (0);
    }
    show = YAML;
  }
  else
  if(argc >= 2)
  {
    int pos = 1, i;
    char *wrong_arg = 0;
    while(pos < argc)
    {
      switch(argv[pos][0])
      {
        case '-':
            for(i = 1; pos < argc && i < (int)strlen(argv[pos]); ++i)
            switch (argv[pos][i])
            {
              case 'i': OY_PARSE_STRING_ARG(input_file_name); break;
              case 'x': OY_PARSE_STRING_ARG(xpath); break;
              case 'c': show = COUNT; break;
              case 'j': show = JSON; break;
              case 'k': show = KEY; break;
              case 'p': show = PATHS; break;
              case 's': OY_PARSE_STRING_ARG(value_string); break;
              case 't': show = TYPE; break;
              case 'v': ++verbose; break;
              case 'y': show = YAML; break;
              case 'h':
              case '-':
                        if(i == 1)
                        {
                             if(OY_IS_ARG("verbose"))
                        { ++verbose; i=100; break; }
                        } OYJL_FALLTHROUGH
              default:
                        printfHelp(argc, argv);
                        exit (0);
                        break;
            }
            break;
        default:
                        printfHelp(argc, argv);
                        exit (0);
                        break;
      }
      if( wrong_arg )
      {
       fprintf( stderr, "%s %s", "wrong argument to option:", wrong_arg);
       printfHelp(argc, argv);
       exit(1);
      }
      ++pos;
    }
  } else
  {
                        printfHelp(argc, argv);
                        exit (0);
  }

  if(input_file_name)
  {
    FILE * fp;
   
    if(strcmp(input_file_name,"-") == 0)
      fp = stdin;
    else
      fp = fopen(input_file_name,"rb");

    if(fp)
    {
      text = oyjlReadFileStreamToMem( fp, &size ); 
      if(fp != stdin) fclose( fp );
    }
  }

  if(text)
  {
    char error_buffer[256] = {0};
    if(verbose)
      fprintf(stderr, "file read:\t\"%s\"\n", input_file_name);
    root = oyjlTreeParse( text, error_buffer, 256 );
    if(error_buffer[0] != '\000')
      fprintf(stderr, "ERROR:\t\"%s\"\n", error_buffer);
    if(verbose)
      fprintf(stderr, "file parsed:\t\"%s\"\n", input_file_name);

    if(xpath)
    {
      char ** paths = NULL;
      int count = 0, i;

      oyjlTreeToPaths( root, 1000000, xpath, 0, &paths );
      while(paths && paths[count]) ++count;

      if(show == PATHS)
        for(i = 0; i < count; ++i)
          fprintf(stdout,"%s\n", paths[i]);
      else if(show == KEY)
        fprintf(stdout,"%s\n", (count && paths[0] && strlen(strchr(paths[0],'/'))) ? strrchr(paths[0],'/') + 1 : "");

      if(paths || value_string)
        value = oyjlTreeGetValue( root,
                                     value_string ? OYJL_CREATE_NEW : 0,
                                     paths?paths[0]:xpath );
      if(verbose)
        fprintf(stderr, "%s xpath \"%s\"\n", value?"found":"found not", xpath);

      oyjlStringListRelease( &paths, count, free );

      if(value_string)
      {
        if(value)
          oyjlValueSetString( value, value_string );
        else
          oyjlMessage_p( oyjlMSG_ERROR, 0, OYJL_DBG_FORMAT_"obtained no leave for xpath \"%s\" from JSON:\t\"%s\"",
                         OYJL_DBG_ARGS_, xpath, input_file_name );
      }

      if(verbose)
        fprintf(stderr, "processed:\t\"%s\"\n", input_file_name);
    }
    else if(value_string)
      oyjlMessage_p( oyjlMSG_ERROR, 0, OYJL_DBG_FORMAT_"set argument needs as well a xpath argument", OYJL_DBG_ARGS_ );
    else
      value = root;
  }

  if(value)
  switch(show)
  {
    case JSON:
    case YAML:
      {
        char * text = NULL;
        int level = 0;
        if(show == JSON)
          oyjlTreeToJson( value_string ? root : value, &level, &text );
        else if(show == YAML)
        {
          oyjlTreeToYaml( value_string ? root : value, &level, &text );
          oyjlStringAdd( &text, 0,0, "\n" );
        }
        if(text)
        {
          fwrite( text, sizeof(char), strlen(text), stdout );
          free(text);
        }
      }
      break;
    case TYPE:
      switch(value->type)
      {
        case oyjl_t_string: puts( "string" ); break;
        case oyjl_t_number: puts( "number" ); break;
        case oyjl_t_object: puts( "object" ); break;
        case oyjl_t_array: puts( "array" ); break;
        case oyjl_t_true: puts( "true" ); break;
        case oyjl_t_false: puts( "false" ); break;
        case oyjl_t_null: puts( "null" ); break;
        case oyjl_t_any: puts( "any" ); break;
      }
      break;
    case COUNT:
      {
        char n[128] = {0};
        sprintf(n, "%d", oyjlValueCount(value));
        puts( n );
      }
      break;
    case KEY:
      if(!xpath && value->type == oyjl_t_object && oyjlValueCount(value) > index)
        puts( value->u.object.keys[index] );
      break;
    case PATHS:
      if(!xpath)
      {
        char ** paths = NULL;
        int count = 0, i;

        oyjlTreeToPaths( root, 1000000, NULL, 0, &paths );
        if(verbose)
          fprintf(stderr, "processed:\t\"%s\"\n", input_file_name);
        while(paths && paths[count]) ++count;

        for(i = 0; i < count; ++i)
          fprintf(stdout,"%s\n", paths[i]);

        oyjlStringListRelease( &paths, count, free );
      }
      break;
    default: break;
  }

  if(root) oyjlTreeFree( root );
  if(text) free(text);

  return error;
}

